File: app/src/main/java/com/jobtick/fragments/MakeAnOfferBudgetFragment.java
package com.jobtick.fragments;

import android.os.Bundle;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.core.content.ContextCompat;
import androidx.fragment.app.Fragment;

import com.android.volley.AuthFailureError;
import com.android.volley.DefaultRetryPolicy;
import com.android.volley.NetworkResponse;
import com.android.volley.Request;
import com.android.volley.RequestQueue;
import com.android.volley.toolbox.StringRequest;
import com.android.volley.toolbox.Volley;
import com.google.android.material.button.MaterialButton;
import com.google.gson.Gson;
import com.jobtick.R;
import android.annotation.SuppressLint;

import com.jobtick.activities.ActivityBase;
import com.jobtick.activities.MakeAnOfferActivity;
import com.jobtick.activities.TaskDetailsActivity;
import com.jobtick.models.MakeAnOfferModel;
import com.jobtick.models.TaskModel;
import com.jobtick.models.UserAccountModel;
import com.jobtick.models.calculation.EarningCalculationModel;
import com.jobtick.utils.Constant;
import com.jobtick.utils.ConstantKey;
import com.jobtick.utils.SessionManager;
import com.jobtick.widget.ExtendedEntryText;

import org.json.JSONException;
import org.json.JSONObject;

import java.util.HashMap;
import java.util.Locale;
import java.util.Map;

import butterknife.BindView;
import butterknife.ButterKnife;

/**
 * A simple {@link Fragment} subclass.
 */
public class MakeAnOfferBudgetFragment extends Fragment implements TextWatcher {

    @SuppressLint("NonConstantResourceId")
    @BindView(R.id.edt_budget)
    ExtendedEntryText edtBudget;
    @SuppressLint("NonConstantResourceId")
    @BindView(R.id.txt_service_fee)
    TextView txtServiceFee;
    @SuppressLint("NonConstantResourceId")
    @BindView(R.id.txt_final_budget)
    TextView txtFinalBudget;
    @SuppressLint("NonConstantResourceId")
    @BindView(R.id.txt_account_level)
    TextView txtAccountLevel;
    @SuppressLint("NonConstantResourceId")
    @BindView(R.id.txt_current_service_fee)
    TextView txtCurrentServiceFee;
    @SuppressLint("NonConstantResourceId")
    @BindView(R.id.txt_learn_how_level_affects_service_fee)
    TextView txtLearnHowLevelAffectsServiceFee;
    @SuppressLint("NonConstantResourceId")
    @BindView(R.id.btn_next)
    MaterialButton btnNext;
    @SuppressLint("NonConstantResourceId")
    @BindView(R.id.img_level)
    ImageView imgLevel;
    @SuppressLint("NonConstantResourceId")
    @BindView(R.id.tvOffer)
    TextView tvOffer;

    Gson gson;

    private MakeAnOfferModel makeAnOfferModel;
    private MakeAnOfferActivity makeAnOfferActivity;
    BudgetCallbackFunction budgetCallbackFunction;
    private UserAccountModel userAccountModel;
    private TaskModel taskModel;
    private SessionManager sessionManager;

    public static MakeAnOfferBudgetFragment newInstance(MakeAnOfferModel makeAnOfferModel, BudgetCallbackFunction budgetCallbackFunction) {

        Bundle args = new Bundle();
        args.putParcelable(ConstantKey.MAKE_AN_OFFER_MODEL, makeAnOfferModel);
        // args.putParcelable(ConstantKey.TASK, taskModel);
        MakeAnOfferBudgetFragment fragment = new MakeAnOfferBudgetFragment();
        fragment.budgetCallbackFunction = budgetCallbackFunction;
        fragment.setArguments(args);
        return fragment;
    }

    public MakeAnOfferBudgetFragment() {
        // Required empty public constructor
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        View view = inflater.inflate(R.layout.fragment_make_an_offer_budget, container, false);
        ButterKnife.bind(this, view);
        gson = new Gson();
        return view;
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        makeAnOfferActivity = (MakeAnOfferActivity) getActivity();
        if (makeAnOfferActivity != null) {
            sessionManager = new SessionManager(makeAnOfferActivity);
        }
        makeAnOfferModel = new MakeAnOfferModel();
        userAccountModel = new UserAccountModel();
        userAccountModel = sessionManager.getUserAccount();
        if (getArguments() != null && getArguments().getParcelable(ConstantKey.MAKE_AN_OFFER_MODEL) != null) {
            makeAnOfferModel = getArguments().getParcelable(ConstantKey.MAKE_AN_OFFER_MODEL);
        }
        /*if (getArguments() != null && getArguments().getParcelable(ConstantKey.TASK) != null) {
            taskModel = getArguments().getParcelable(ConstantKey.TASK);
        }*/
        taskModel = TaskDetailsActivity.taskModel;
        if (makeAnOfferModel != null) {
            initLayout();
        }
        //toolbar.setNavigationOnClickListener(MakeAnOfferBudgetFragment.this);

        //Custom tool bar back here
        ImageView ivBack = (ImageView) view.findViewById(R.id.ivBack);
        ivBack.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {

                makeAnOfferActivity.onBackPressed();
            }
        });
        edtBudget.addTextChangedListener(this);

        btnNext.setOnClickListener(v -> {
            if (budgetCallbackFunction != null) {
                if (!validation(true)) return;

                makeAnOfferModel.setOffer_price(Integer.parseInt(edtBudget.getText().trim()));
                budgetCallbackFunction.continueButtonBudget(makeAnOfferModel);
            }
        });
    }

    private void initLayout() {
        //edtBudget.setText(String.format("%d", makeAnOfferModel.getOffer_price()));
        tvOffer.setText(String.format(Locale.ENGLISH, "$%d", taskModel.getBudget()));
        txtAccountLevel.setText(String.format(Locale.ENGLISH, "Level %d", userAccountModel.getWorkerTier().getId()));

        if (userAccountModel.getWorkerTier().getId() == 1) {
            //TODO change image level 1
            imgLevel.setImageDrawable(ContextCompat.getDrawable(requireContext(), R.drawable.ic_medal1));
        } else if (userAccountModel.getWorkerTier().getId() == 2) {
            //TODO change image level 2
            imgLevel.setImageDrawable(ContextCompat.getDrawable(requireContext(), R.drawable.ic_medal2));
        } else if (userAccountModel.getWorkerTier().getId() == 3) {
            //TODO change image level 3
            imgLevel.setImageDrawable(ContextCompat.getDrawable(requireContext(), R.drawable.ic_medal3));
        } else if (userAccountModel.getWorkerTier().getId() == 4) {
            //TODO change image level 4
            imgLevel.setImageDrawable(ContextCompat.getDrawable(requireContext(), R.drawable.ic_medal4));
        }
    }

    @Override
    public void beforeTextChanged(CharSequence charSequence, int i, int i1, int i2) {

    }

    @Override
    public void onTextChanged(CharSequence charSequence, int i, int i1, int i2) {

    }

    @Override
    public void afterTextChanged(Editable editable) {
        btnNext.setEnabled(true);

        if (!validation(false)) {
            txtServiceFee.setText(R.string.dollar_00);
            txtFinalBudget.setText(R.string.dollar_00);
            txtCurrentServiceFee.setText("0%");
            return;
        }
        calculate(editable.toString());
    }

    private boolean validation(boolean showToast) {
        if (edtBudget.getText().length() == 0) {
            if (showToast)
                ((ActivityBase) requireActivity()).showToast("Please enter your offer.", requireContext());
            return false;
        }
        if (Integer.parseInt(edtBudget.getText()) < 5) {
            if (showToast)
                ((ActivityBase) requireActivity()).showToast("You can't offer lower than 5 dollars.", requireContext());
            return false;
        }
        if (Integer.parseInt(edtBudget.getText()) < taskModel.getBudget()) {
            if (showToast)
                ((ActivityBase) requireActivity()).showToast("You can't offer lower than poster offer.", requireContext());
            return false;
        }
        if (Integer.parseInt(edtBudget.getText()) > 9999) {
            if (showToast)
                ((ActivityBase) requireActivity()).showToast("You can't offer more than 9999 dollars.", requireContext());
            return false;
        }
        return true;
    }


    public void calculate(String amount) {

        StringRequest stringRequest = new StringRequest(Request.Method.POST, Constant.URL_OFFERS_EARNING_CALCULATION,
                response -> {
                    try {
                        JSONObject jsonObject = new JSONObject(response);
                        String data = jsonObject.getString("data");

                        EarningCalculationModel model = gson.fromJson(data, EarningCalculationModel.class);
                        txtServiceFee.setText(String.format(Locale.ENGLISH, "$%.1f", (model.getServiceFee() + model.getGstAmount())));
                        txtFinalBudget.setText(String.format(Locale.ENGLISH, "$%.1f", model.getNetEarning()));
                        txtCurrentServiceFee.setText(String.format(Locale.ENGLISH, "%s", model.getServiceFeeRate()));

                    } catch (JSONException e) {
                        e.printStackTrace();
                    }
                },
                error -> {
                    NetworkResponse networkResponse = error.networkResponse;
                    if (networkResponse != null && networkResponse.data != null) {
                        String jsonError = new String(networkResponse.data);
                        try {
                            JSONObject jsonObject = new JSONObject(jsonError);
                            JSONObject jsonObject_error = jsonObject.getJSONObject("error");
                            String message = jsonObject_error.getString("message");
                            ((ActivityBase) requireActivity()).showToast(message, requireContext());

                        } catch (JSONException e) {
                            e.printStackTrace();
                            ((ActivityBase) requireActivity()).showToast("Something went wrong", requireContext());
                        }
                    } else {
                        ((ActivityBase) requireActivity()).showToast("Something went wrong", requireContext());
                    }

                }) {

            @Override
            public Map<String, String> getHeaders() throws AuthFailureError {
                Map<String, String> map1 = new HashMap<String, String>();
                map1.put("authorization", sessionManager.getTokenType() + " " + sessionManager.getAccessToken());
                map1.put("Content-Type", "application/x-www-form-urlencoded");
                map1.put("X-Requested-With", "XMLHttpRequest");

                return map1;
            }

            @Override
            protected Map<String, String> getParams() {
                Map<String, String> map1 = new HashMap<String, String>();
                map1.put("amount", amount);
                return map1;
            }
        };

        stringRequest.setRetryPolicy(new DefaultRetryPolicy(0, -1,
                DefaultRetryPolicy.DEFAULT_BACKOFF_MULT));
        RequestQueue requestQueue = Volley.newRequestQueue(requireContext());
        requestQueue.add(stringRequest);
    }


    public interface BudgetCallbackFunction {

        void continueButtonBudget(MakeAnOfferModel makeAnOfferModel);
    }
}

File: app/src/main/java/com/jobtick/android/fragments/MakeAnOfferBudgetFragment.kt
package com.jobtick.android.fragments

import android.os.Bundle
import android.os.Parcelable
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.appcompat.content.res.AppCompatResources
import androidx.appcompat.widget.AppCompatImageView
import androidx.core.widget.doOnTextChanged
import androidx.fragment.app.Fragment
import butterknife.ButterKnife
import com.android.volley.*
import com.android.volley.toolbox.StringRequest
import com.android.volley.toolbox.Volley
import com.google.android.material.button.MaterialButton
import com.google.android.material.textfield.TextInputLayout
import com.google.android.material.textview.MaterialTextView
import com.google.gson.Gson
import com.jobtick.android.BuildConfig
import com.jobtick.android.R
import com.jobtick.android.activities.ActivityBase
import com.jobtick.android.activities.MakeAnOfferActivity
import com.jobtick.android.models.MakeAnOfferModel
import com.jobtick.android.models.TaskModel
import com.jobtick.android.models.UserAccountModel
import com.jobtick.android.models.calculation.EarningCalculationModel
import com.jobtick.android.utils.Constant
import com.jobtick.android.utils.ConstantKey
import com.jobtick.android.utils.SessionManager
import org.json.JSONException
import org.json.JSONObject
import java.util.*

/**
 * A simple [Fragment] subclass.
 */
class MakeAnOfferBudgetFragment : Fragment() {
    private lateinit var txtServiceFee: TextView
    private lateinit var txtFeeTitle: TextView
    private lateinit var txtFinalBudget: TextView
    private lateinit var txtLearnHowLevelAffectsServiceFee: TextView
    private lateinit var btnNext: MaterialButton
    private lateinit var tvOffer: TextView
    private var gson: Gson? = null
    private var makeAnOfferModel: MakeAnOfferModel? = null
    private var makeAnOfferActivity: MakeAnOfferActivity? = null
    private var budgetCallbackFunction: BudgetCallbackFunction? = null
    private var userAccountModel: UserAccountModel? = null
    private var taskModel: TaskModel? = null
    private var sessionManager: SessionManager? = null
    private lateinit var budget: TextInputLayout
    private lateinit var infoIcon: AppCompatImageView
    private lateinit var label: MaterialTextView

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        // Inflate the layout for this fragment
        val view = inflater.inflate(R.layout.fragment_make_an_offer_budget, container, false)
        ButterKnife.bind(this, view)
        gson = Gson()
        return view
    }

    override fun onResume() {
        super.onResume()
        if (budget.editText!!.text.isNotEmpty()) {
            btnNext.isEnabled = true
        }
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        setIds()
        makeAnOfferActivity = requireActivity() as MakeAnOfferActivity
        sessionManager = SessionManager(makeAnOfferActivity)

        makeAnOfferModel = MakeAnOfferModel()
        userAccountModel = UserAccountModel()
        userAccountModel = sessionManager!!.userAccount
        if (arguments != null && requireArguments().getParcelable<Parcelable?>(ConstantKey.MAKE_AN_OFFER_MODEL) != null) {
            makeAnOfferModel = requireArguments().getParcelable(ConstantKey.MAKE_AN_OFFER_MODEL)
        }
        if (arguments != null && requireArguments().getParcelable<Parcelable?>(ConstantKey.TASK) != null) {
            taskModel = requireArguments().getParcelable(ConstantKey.TASK)
        }
        if (makeAnOfferModel != null) {
            initLayout()
        }
        //toolbar.setNavigationOnClickListener(MakeAnOfferBudgetFragment.this);

        //Custom tool bar back here
        val ivBack = view.findViewById<ImageView>(R.id.ivBack)
        ivBack.setOnClickListener { makeAnOfferActivity!!.onBackPressed() }
        btnNext.setOnClickListener {
            if (budgetCallbackFunction != null) {
                if (!validation(true)) return@setOnClickListener
                makeAnOfferModel!!.offer_price = budget.editText!!.text.trim { it <= ' ' }.toString().toInt()
                budgetCallbackFunction!!.continueButtonBudget(makeAnOfferModel)
            }
        }
        txtFeeTitle.setOnClickListener { v: View? ->
            val infoBottomSheet = ServiceFeeInfoBottomSheet()
            infoBottomSheet.show(childFragmentManager, null)
        }
        txtLearnHowLevelAffectsServiceFee.setOnClickListener { v: View? -> }
    }

    private fun setIds() {
        txtServiceFee = requireView().findViewById(R.id.txt_service_fee)
        txtFeeTitle = requireView().findViewById(R.id.txt_fee_title)
        txtFinalBudget = requireView().findViewById(R.id.txt_final_budget)
        tvOffer = requireView().findViewById(R.id.tvOffer)
        txtLearnHowLevelAffectsServiceFee = requireView().findViewById(R.id.txt_learn_how_level_affects_service_fee)
        btnNext = requireView().findViewById(R.id.btn_next)
        budget = requireView().findViewById(R.id.budget)
        infoIcon = requireView().findViewById(R.id.info)
        label = requireView().findViewById(R.id.label)


        budget.editText!!.setOnFocusChangeListener { _, b ->
            if (b) {
                budget.editText!!.setCompoundDrawablesWithIntrinsicBounds(R.drawable.ic_dolar, 0, 0, 0)
            }
            if (b) {
                infoIcon.setColorFilter(resources.getColor(R.color.primary_500), android.graphics.PorterDuff.Mode.SRC_IN)
                label.setTextColor(resources.getColor(R.color.neutral_light_600))
            } else {
                infoIcon.setColorFilter(resources.getColor(R.color.neutral_light_400), android.graphics.PorterDuff.Mode.SRC_IN)
                label.setTextColor(resources.getColor(R.color.neutral_light_400))
            }
        }
        budget.editText?.doOnTextChanged { text, _, _, _ ->
            btnNext.isEnabled = validation(false)
            if (validation(false)) {
                infoIcon.setColorFilter(resources.getColor(R.color.primary_500), android.graphics.PorterDuff.Mode.SRC_IN)
                label.setTextColor(resources.getColor(R.color.neutral_light_600))
                budget.boxStrokeColor = resources.getColor(R.color.primary_500)
                budget.hintTextColor = AppCompatResources.getColorStateList(requireContext(), R.color.primary_500)
            } else {
                infoIcon.setColorFilter(resources.getColor(R.color.primary_error), android.graphics.PorterDuff.Mode.SRC_IN)
                label.setTextColor(resources.getColor(R.color.primary_error))
                budget.boxStrokeColor = resources.getColor(R.color.primary_error)
                budget.hintTextColor = AppCompatResources.getColorStateList(requireContext(), R.color.primary_error)
            }
            text?.let {
                if (it.isNotEmpty()) calculate(it.toString())
            }
            if (text!!.isNotEmpty())
                budget.editText!!.setCompoundDrawablesWithIntrinsicBounds(R.drawable.ic_dolar, 0, 0, 0)
            else
                budget.editText!!.setCompoundDrawablesWithIntrinsicBounds(0, 0, 0, 0)
        }
    }

    private fun initLayout() {
        if (makeAnOfferModel != null && makeAnOfferModel!!.offer_price != 0) {
            budget.editText!!.setText(String.format(Locale.ENGLISH, "%d", makeAnOfferModel!!.offer_price))
            txtServiceFee.text = String.format(Locale.ENGLISH, "$%.1f", makeAnOfferModel!!.allFee)
            txtFinalBudget.text = String.format(Locale.ENGLISH, "$%.1f", makeAnOfferModel!!.youWillReceive)
        }
        tvOffer.text = String.format(Locale.ENGLISH, "$%d", taskModel!!.budget)
    }


    var is4Digits = false


    private fun validation(showToast: Boolean): Boolean {
        if (budget.editText!!.text.isEmpty()) {
            if (showToast) (requireActivity() as ActivityBase).showToast("Please enter your offer.", requireContext())
            return false
        }
        if (budget.editText!!.text.toString().toInt() < 5) {
            if (showToast) (requireActivity() as ActivityBase).showToast("You can't offer lower than 5 dollars.", requireContext())
            return false
        }
        if (budget.editText!!.text.toString().toInt() < taskModel!!.budget) {
            if (showToast) (requireActivity() as ActivityBase).showToast("You can't offer lower than poster offer.", requireContext())
            return false
        }
        if (budget.editText!!.text.toString().toInt() > 9999) {
            if (showToast) (requireActivity() as ActivityBase).showToast("You can't offer more than 9999 dollars.", requireContext())
            return false
        }
        return true
    }

    fun calculate(amount: String) {
        val stringRequest: StringRequest = object : StringRequest(Method.POST, Constant.URL_OFFERS_EARNING_CALCULATION,
                Response.Listener { response: String? ->
                    try {
                        val jsonObject = JSONObject(response)
                        val data = jsonObject.getString("data")
                        val model = gson!!.fromJson(data, EarningCalculationModel::class.java)
                        makeAnOfferModel!!.allFee = model.serviceFee + model.gstAmount
                        makeAnOfferModel!!.youWillReceive = model.netEarning
                        txtServiceFee.text = String.format(Locale.ENGLISH, "$%.1f", makeAnOfferModel!!.allFee)
                        txtFinalBudget.text = String.format(Locale.ENGLISH, "$%.1f", makeAnOfferModel!!.youWillReceive)
                    } catch (e: JSONException) {
                        e.printStackTrace()
                    }
                },
                Response.ErrorListener { error: VolleyError ->
                    val networkResponse = error.networkResponse
                    if (networkResponse?.data != null) {
                        val jsonError = String(networkResponse.data)
                        try {
                            val jsonObject = JSONObject(jsonError)
                            val jsonObjectError = jsonObject.getJSONObject("error")
                            val message = jsonObjectError.getString("message")
                            (requireActivity() as ActivityBase).showToast(message, requireContext())
                        } catch (e: JSONException) {
                            e.printStackTrace()
                            (requireActivity() as ActivityBase).showToast("Something went wrong", requireContext())
                        }
                    } else {
                        (requireActivity() as ActivityBase).showToast("Something went wrong", requireContext())
                    }
                }) {
            @Throws(AuthFailureError::class)
            override fun getHeaders(): Map<String, String> {
                val map1: MutableMap<String, String> = HashMap()
                map1["authorization"] = sessionManager!!.tokenType + " " + sessionManager!!.accessToken
                map1["Content-Type"] = "application/x-www-form-urlencoded"
                map1["X-Requested-With"] = "XMLHttpRequest"
                map1["Version"] = BuildConfig.VERSION_CODE.toString()
                return map1
            }

            override fun getParams(): Map<String, String> {
                val map1: MutableMap<String, String> = HashMap()
                map1["amount"] = amount
                return map1
            }
        }
        stringRequest.retryPolicy = DefaultRetryPolicy(0, -1,
                DefaultRetryPolicy.DEFAULT_BACKOFF_MULT)
        val requestQueue = Volley.newRequestQueue(requireContext())
        requestQueue.add(stringRequest)
    }

    interface BudgetCallbackFunction {
        fun continueButtonBudget(makeAnOfferModel: MakeAnOfferModel?)
    }

    companion object {
        fun newInstance(makeAnOfferModel: MakeAnOfferModel?, taskModel: TaskModel?, budgetCallbackFunction: BudgetCallbackFunction?): MakeAnOfferBudgetFragment {
            val args = Bundle()
            args.putParcelable(ConstantKey.MAKE_AN_OFFER_MODEL, makeAnOfferModel)
            args.putParcelable(ConstantKey.TASK, taskModel)
            val fragment = MakeAnOfferBudgetFragment()
            fragment.budgetCallbackFunction = budgetCallbackFunction
            fragment.arguments = args
            return fragment
        }
    }
}